### PR TITLE
`azurerm_kubernetes_cluster_node_pool` - `os_sku` supports update

### PR DESCRIPTION
**Pull request by ms-henglu**
## Community Note

* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave "+1" or "me too" comments, they generate extra noise for PR followers and do not help prioritize for review

## PR Checklist

- [ ] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [ ] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [ ] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [ ] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [ ] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”

## Changes to existing Resource / Data Source

- [ ] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [ ] I have written new tests for my resource or datasource changes & updated any relevent documentation.
- [ ] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.
- [ ] (For changes that include a **state migration only**). I have manually tested the migration path between relevant versions of the provider.

## Testing 

- [ ] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

```
=== RUN   TestAccKubernetesClusterNodePool_osSkuMigration
=== PAUSE TestAccKubernetesClusterNodePool_osSkuMigration
=== CONT  TestAccKubernetesClusterNodePool_osSkuMigration
--- PASS: TestAccKubernetesClusterNodePool_osSkuMigration (1066.68s)
PASS
```

## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

* `azurerm_resource` - support for the `thing1` property [GH-00000]

This is a (please select all that apply):

- [ ] Bug Fix
- [ ] New Feature (ie adding a service, resource, or data source)
- [ ] Enhancement
- [ ] Breaking Change

## Related Issue(s)
Fixes #0000

> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

### CODE REVIEW

**Review by stephybun — COMMENTED** [review]
Thanks @ms-henglu but given the ongoing discussions about whether to support preview features in the AKS resource or not we can't progress with this since this may potentially need to be removed again in the next major version of the provider.

Could we close this PR for the moment, we're happy to reopen once we have a resolution on the discussion.

**Comment by ms-henglu** [comment]
Agreed! I'll close this task.

File: website/docs/r/kubernetes_cluster_node_pool.html.markdown [hunk 1]
```diff
@@ -126,7 +126,9 @@ The following arguments are supported:
...
 
-* `os_sku` - (Optional) Specifies the OS SKU used by the agent pool. Possible values are `AzureLinux`, `Ubuntu`, `Windows2019` and `Windows2022`. If not specified, the default is `Ubuntu` if OSType=Linux or `Windows2019` if OSType=Windows. And the default Windows OSSKU will be changed to `Windows2022` after Windows2019 is deprecated. Changing this forces a new resource to be created.
+* `os_sku` - (Optional) Specifies the OS SKU used by the agent pool. Possible values are `AzureLinux`, `Ubuntu`, `Windows2019` and `Windows2022`. If not specified, the default is `Ubuntu` if OSType=Linux or `Windows2019` if OSType=Windows. And the default Windows OSSKU will be changed to `Windows2022` after Windows2019 is deprecated. Changing this from `AzureLinux` or `Ubuntu` to `AzureLinux` or `Ubuntu` will not replace the resource, otherwise it forces a new resource to be created.
+
+-> **Note:**  This requires that the Preview Feature `Microsoft.ContainerService/OSSKUMigrationPreview` is enabled and the Resource Provider is re-registered, see [the documentation](https://learn.microsoft.com/en-us/azure/azure-linux/tutorial-azure-linux-migration) for more information.
```
Comment (stephybun) [inline website/docs/r/kubernetes_cluster_node_pool.html.markdown:131]: This should be removed if it's not really a preview feature anymore
```suggestion
```

**Review by stephybun — CHANGES_REQUESTED** [review]

File: website/docs/r/kubernetes_cluster_node_pool.html.markdown (hunk 1 above)
Comment (ms-henglu) [reply website/docs/r/kubernetes_cluster_node_pool.html.markdown:131]: I'll remove it when it's GA, currently we still need this feature registered.

Comment (stephybun) [reply website/docs/r/kubernetes_cluster_node_pool.html.markdown:131]: Does it though? That flag isn't registered on our testing subscription
```
$ az feature show --namespace "Microsoft.ContainerService" --name "OSSKUMigrationPreview"
{
  "id": "/subscriptions/*/providers/Microsoft.Features/providers/microsoft.ContainerService/features/OSSKUMigrationPreview",
  "name": "microsoft.ContainerService/OSSKUMigrationPreview",
  "properties": {
    "state": "NotRegistered"
  },
  "type": "Microsoft.Features/providers/features"
}
```

And the test passes
image

Comment (ms-henglu) [reply website/docs/r/kubernetes_cluster_node_pool.html.markdown]: Got it, I'll remove the notice.

**Review by hbeberman — CHANGES_REQUESTED** [review]
Left a few comments on things that need to change.

Additionally the function `resourceKubernetesClusterNodePoolUpdate` in [kubernetes_cluster_node_pool_resource.go](https://github.com/hashicorp/terraform-provider-azurerm/blob/b4185b4899f93d624103c9f7f5d7b9e994ed0595/internal/services/containers/kubernetes_cluster_node_pool_resource.go#L745) will need to apply the os_sku if it's changed.

```
	if d.HasChange("os_sku") {
		if osSku := d.Get("os_sku").(string); osSku != "" {
			props.OsSKU = pointer.To(agentpools.OSSKU(osSku))
		}
	}
```

File: internal/services/containers/kubernetes_cluster_node_pool_resource.go [hunk 2]
```diff
@@ -269,7 +281,6 @@ func resourceKubernetesClusterNodePoolSchema() map[string]*pluginsdk.Schema {
 		"os_sku": {
 			Type:     pluginsdk.TypeString,
 			Optional: true,
-			ForceNew: true,
```
[the remark this answers is not in the document]
Comment (ms-henglu) [reply internal/services/containers/kubernetes_cluster_node_pool_resource.go:272]: Yes, I opened a separate PR for `kubernetes_cluster_resource`

**Comment by ms-henglu** [comment]
Thanks @hbeberman for the code review. I've updated this PR as suggested.

**Review by hbeberman — APPROVED** [review]

**Comment by ms-henglu** [comment]
Hi @stephybun , the tests were passing because there's a bug in my PR that the `forceNew:true` is not removed, so the node pool is replaced.

I've fixed it, and the tests are failing because the preview feature flag is not enabled:
```
=== RUN   TestAccKubernetesClusterNodePool_osSkuMigration
=== PAUSE TestAccKubernetesClusterNodePool_osSkuMigration
=== CONT  TestAccKubernetesClusterNodePool_osSkuMigration
    testcase.go:113: Step 3/6 error: Error running apply: exit status 1
        Error: updating Node Pool Agent Pool (Subscription: "*******"
        Resource Group Name: "acctestRG-aks-240628111457303318"
        Managed Cluster Name: "acctestaks240628111457303318"
        Agent Pool Name: "internal"): performing CreateOrUpdate: unexpected status 400 (400 Bad Request) with response: {
          "code": "PropertyChangeNotAllowed",
          "details": null,
          "message": "Changing property 'agentPoolProfile.OSSKU' is not allowed.",
          "subcode": "",
          "target": "agentPoolProfile.OSSKU"
         }
          with azurerm_kubernetes_cluster_node_pool.test,
          on terraform_plugin_test.tf line 45, in resource "azurerm_kubernetes_cluster_node_pool" "test":
          45: resource "azurerm_kubernetes_cluster_node_pool" "test" {
```

**Comment by ms-henglu** [comment]
Hi @stephybun - I also confirmed with the service team that this feature flag is required. I've reverted the previous commit which removes the notice.

**Review by stephybun — APPROVED** [review]
Thanks @ms-henglu LGTM 👍

**Pull request merged**